src/docs/introduction/upgrading-to-stencil-three.md

```diff
@@ -0,0 +1,258 @@
+```ts
+// stencil.config.ts
+import { Config } from '@stencil/core';
+
+export const config: Config = {
+  outputTargets: [
+    {
+      type: 'dist-custom-elements',
+      generateTypeDeclarations: false,
+      // ...
+    },
+    // ...
+  ],
+  // ...
+};
+```
+
+### Deprecated `assetsDir` Removed from `@Component()` decorator
+The `assetsDir` field was [deprecated in Stencil v2.0.0](https://github.com/ionic-team/stencil/blob/main/BREAKING_CHANGES.md#componentassetsdir), but some backwards compatibility was retained with a warning message.
+It has been fully removed in Stencil v3.0.0 in favor of `assetsDirs`.
+To migrate from existing usages of `assetsDir`, update the property name and wrap its value in an array:
+```diff
+@Component({
+  tag: 'my-component',
+- assetsDir: 'assets',
++ assetsDirs: ['assets'],
+})
+```
+For more information on the `assetsDirs` field, please see the [Stencil Documentation on `assetsDirs`](/docs/assets#assetsdirs)
+
+### Drop Node 12 Support
+Stencil no longer supports Node 12.
+Please upgrade local development machines, continuous integration pipelines, etc. to use Node v14 or higher.
+
+### Strongly Typed Inputs
+`onInput` and `onInputCapture` events have had their interface's updated to accept an argument of `InputEvent` over `Event`:
+```diff
+- onInput?: (event: Event) => void;
++ onInput?: (event: InputEvent) => void;
+- onInputCapture?: (event: Event) => void;
++ onInputCapture?: (event: InputEvent) => void;
+```
+`event` arguments to either callback should be updated to take this narrower typing into account
+
+### Narrowed Typing for `autocapitalize` Attribute
+The [`autocaptialize` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autocapitalize) has been narrowed from type `any` to type `string`.
+This changes brings Stencil into closer alignment with TypeScript's typings for the attribute.
+No explicit changes are needed, unless a project was passing non-strings to the attribute.
+
+### Custom Types for Props and Events are now Exported from `components.d.ts`
+
+Custom types for props and custom events are now re-exported from a project's `components.d.ts` file.
+
+For the following Stencil component
+```tsx
+import { Component, Event, EventEmitter, Prop, h } from '@stencil/core';
+
+export type NameType = string;
+export type Todo = Event;
+
+@Component({
+  tag: 'my-component',
+  styleUrl: 'my-component.css',
+  shadow: true,
+})
+export class MyComponent {
+  @Prop() first: NameType;
+
+  @Event() todoCompleted: EventEmitter<Todo>
+
+  render() {
+    return <div>Hello, World! I'm {this.first}</div>;
+  }
+}
+```
+
+The following data will now be included automatically in `components.d.ts`:
+```diff
+ import { HTMLStencilElement, JSXBase } from "@stencil/core/internal";
+ import { NameType } from "./components/my-component/my-component";
++export { NameType } from "./components/my-component/my-component";
```

Comment:
We should include the events too if we're calling out how to import them later in this section
```suggestion
 import { NameType, Todo } from "./components/my-component/my-component";
+export { NameType, Todo } from "./components/my-component/my-component";
```

Reply:
I don't necessarily disagree, but this is the output of `npm run build` for the component immediately above this `component.d.ts` snippet. I think this should reflect was generated by Stencil

Reply:
alternatively, should we just remove the 'event' stuff since they're exported in 2.X?

Reply:
Really? You should get the `Todo` type in there too. I built the same component and had `Todo` included in these statements

Reply:
Ah, ok. I see what happened. There was a copy of this text without the `export` modifier on `Todo` 🤦